### PR TITLE
Add parity waiting

### PR DESCRIPTION
**Pull request by ogiroux**
This PR adds the free function `cuda::barrier_wait_parity` that takes a pointer to a barrier of any scope and a parity boolean, that we've discussed at length before. It restructures a bit the implementation-internal barrier poll testers to avoid duplication.

### CODE REVIEW

**Comment by ogiroux** [comment]
I'm going to add a `bool try_wait(token)` facility to this PR before it is merged in 2.0.0. Basically this will just go from having it as an internal name, to having it as an external name (remove the `__`).

Discussion from internal developer day pointed out that you can get a good benefit from hiding the latency of the load and/or control the polling backoff yourself. It's already in here, might as well make it usable.

**Comment by wmaxey** [comment]
@ogiroux Does this require any additional tests you would like to see? I was going to permute a few things, but this would be a good opportunity for any edge cases you can think of.

**Comment by wmaxey** [comment]
Virtual: http://scdvs.nvidia.com/Build_Results?virtualId=1000073803&which_page=current_build

**Comment by wmaxey** [comment]
New virtual: http://scdvs.nvidia.com/Build_Results?virtualId=1000073820&which_page=current_build
Test results were clobbered by previous failures from the module_compiler change.

**Comment by wmaxey** [comment]
Latest clean virtual: https://builds4u.nvidia.com/dvs/#/change/3023208755719989.5?eventType=Virtual

File: docs/extended_api/synchronization_primitives/barrier.md [hunk 1]
```diff
@@ -21,8 +21,9 @@ It has the same interface and semantics as [`cuda::std::barrier`], with the
 
 ## Barrier Operations
 
-| [`cuda::barrier::init`]                 | Initialize a `cuda::barrier`. `(friend function)`                 |
-| [`cuda::device::barrier_native_handle`] | Get the native handle to a `cuda::barrier`. `(function template)` |
+| [`cuda::barrier::init`]                        | Initialize a `cuda::barrier`. `(friend function)`                 |
+| [`cuda::device::barrier_native_handle`]        | Get the native handle to a `cuda::barrier`. `(function template)` |
+| [`cuda::barrier::wait_parity/try_wait_parity`] | Wait on a `specific` phase of the barrier |
```
Comment (griwes) [inline docs/extended_api/synchronization_primitives/barrier.md:26]: I don't like this phrasing. @ogiroux how would you feel about "Wait on a phase whose parity matches the argument." or some such?

File: docs/extended_api/synchronization_primitives/barrier/wait_parity.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+Defined in header `<cuda/std/barrier>`:
+
+```cuda
+__host__ __device__ void cuda::std::barrier::wait_parity(bool phase);
+__host__ __device__ bool cuda::std::barrier::try_wait_parity(bool phase);
+```
+
+`barrier::wait_parity` stalls execution while the barrier is not at the specified phase.
+`barrier::try_wait_parity` queries the the state of the barrier against the specified phase.
```
Comment (griwes) [inline docs/extended_api/synchronization_primitives/barrier/wait_parity.md:16]: See the previous comment. It's really weird that the description here doesn't mention parity at all.

**Review by griwes — APPROVED** [review]
LGTM except for the phrasing in the documentation.

**Comment by wmaxey** [comment]
In the interest of time, I'm going to split the documentation changes into another PR and merge the changes.
